include/stratosphere/mitm/mitm_server.hpp

```diff
@@ -21,16 +21,22 @@
 #include "sm_mitm.h"
 #include "mitm_session.hpp"
 
-template <typename T>
+template <typename T, typename... Args>
 class MitmServer : public IWaitable {          
     static_assert(std::is_base_of<IMitmServiceObject, T>::value, "MitM Service Objects must derive from IMitmServiceObject");
     private:
         Handle port_handle;
         unsigned int max_sessions;
         char mitm_name[9];
     
+        std::tuple<Args...> args;
+
+        template<std::size_t... I>
+        static std::shared_ptr<T> ConstructionDetailHelper(std::shared_ptr<Service> forward_service, u64 client_pid, std::tuple<Args...> &&tuple, std::index_sequence<I...>) {
+            return std::make_shared<T>(forward_service, client_pid, std::forward<Args>(std::get<I>(std::forward<std::tuple<Args...>>(tuple)))...);
```

Comment:
Two points:
* You don't need the `std::forward` before `std::get`: `std::get`is a *perfectly forwarding function* that takes care of preserving the value category (="the number of ampersands"). You only need std::forward when referring to forwarding reference parameters (templated type with a "double-ampersand")
* Note that `std::tuple<Args...>&&` only matches rvalue references, but not lvalue references. If you want to match lvalue references to, you need a separate `std::tuple<Args...>&` overload (or use a forwarding reference, which may not be easily achievable in this case though). It looks like you don't need this here, but it's worth being aware of this. If you want to learn more, search for the difference between "forwarding references" and "rvalue references".

Reply:
On second thought, would the following work for you?

```
template<std::size_t... I>
static std::shared_ptr<T> ConstructionDetailHelper(std::shared_ptr<Service> forward_service, u64 client_pid, Args&&... tuple) {
    return std::make_shared<T>(forward_service, client_pid, std::forward<Args>(args)...);
```

If it does, that neatly solves the second issue about only matching rvalue references, since in this case it's using forwarding references.

Reply:
I'm not sure what you're doing there. Can you make sure that code actually makes sense?

Reply:
Regarding the `std::tuple<Args...>&&`, do you mean that will only match when every element of `Args` is an rvalue reference? Strange, because I compiled with `Args = {LogQueue &}`. I wanted to design this to be able to pass through both lvalue references and rvalue references.

Reply:
That should make sense, yea. It seemed unnecessary to me to use a `std::tuple` here, and instead you can just expand the parameter pack directly. Is there any specific part that you find confusing?

Reply:
Regarding my shortened ConstructionDetailHelper example, disregard my comment - I can see why you *need* to pass in a tuple now. Sorry for the confusion!

Reply:
In addition to the other thread, see https://godbolt.org/z/HziW8p <-- note how this code only compiles when providing the second overload.

> Regarding the std::tuple<Args...>&&, do you mean that will only match when every element of Args is an rvalue reference?

Nope, that's not it - say you have a function `template<typename T> func(std::tuple<T>&&)`. You could call this function in different ways, for example
```
func(std::make_tuple(5));
```
or
```
std::tuple<int> my_tuple;
func(my_tuple);
```
`std::make_tuple(5, 3)` will work, since you're passing an rvalue to the function, so that will match the function declaration, which takes an rvalue reference. The second example uses `my_tuple` (i.e. an lvalue), which can't bind against rvalue references, so the example won't compile.

Compare this to a function `template<typename T> other_func(T&&)`. Here, T&& is not an rvalue reference, but a *forwarding reference*. So both function calls `other_func(5);` and `int my_int; other_func(my_int);` will compile.